### PR TITLE
[SPARK-13320] [SQL] Support Star in CreateStruct/CreateArray and Error Handling when DataFrame/DataSet Functions using Star

### PR DESCRIPTION
**Pull request by gatorsmile**
This PR resolves two issues:

First, expanding \* inside aggregate functions of structs when using Dataframe/Dataset APIs. For example, 

``` scala
structDf.groupBy($"a").agg(min(struct($"record.*")))
```

Second, it improves the error messages when having invalid star usage when using Dataframe/Dataset APIs. For example,

``` scala
pagecounts4PartitionsDS
  .map(line => (line._1, line._3))
  .toDF()
  .groupBy($"_1")
  .agg(sum("*") as "sumOccurances")
```

Before the fix, the invalid usage will issue a confusing error message, like:

```
org.apache.spark.sql.AnalysisException: cannot resolve '_1' given input columns _1, _2;
```

After the fix, the message is like:

```
org.apache.spark.sql.AnalysisException: Invalid usage of '*' in function 'sum'
```

cc: @rxin @nongli @cloud-fan

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -362,12 +362,26 @@ class Analyzer(
...
         case e =>
-          e.transformDown {
+          e.transformUp {
+            // ResolveFunctions can handle the case when the number of variables is not valid
             case f1: UnresolvedFunction if containsStar(f1.children) =>
               f1.copy(children = f1.children.flatMap {
                 case s: Star => s.expand(child, resolver)
                 case o => o :: Nil
               })
+            case c: CreateStruct if containsStar(c.children) =>
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Not sure if we have the other functions that can accept `star` as an input parameter. If so, I think we need to create a `trait` for all these case classes. Then, we can remove the duplicate code. Any better idea? Thanks! : )

**Comment by rxin** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -362,12 +362,26 @@ class Analyzer(
...
+              c.copy(children = c.children.flatMap {
+                case s: Star => s.expand(child, resolver)
+                case o => o :: Nil
+              })
+            case c: CreateStructUnsafe if containsStar(c.children) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: `CreateStructUnsafe` only appears after unsafe projection, so I think we don't need to handle it in `Analyzer`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -362,12 +362,26 @@ class Analyzer(
...
+              c.copy(children = c.children.flatMap {
+                case s: Star => s.expand(child, resolver)
+                case o => o :: Nil
+              })
+            // count(*) has been replaced by count(1)
+            case f2: ExpectsInputTypes if containsStar(f2.children) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: why we only fail for `ExpectsInputTypes`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I saw it is being used in two parts in Analyzer. Will remove them. Thanks!

**Comment by cloud-fan** [comment]
The PR title looks confusing, `Star Expansion` is already done, what this PR did is fixing a problem of missing `CreateStruct` when handle stars and adding a better error message, @gatorsmile could you improve it to make it more clear?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: There are 283 expression types, if we exclude 10 `namedExpressions`. Should we cover all these types? Can you help me go over the long list to see if we should include all of them?

**Comment by gatorsmile** [comment]
`Star Expansion` only works when the star are in a `UnresolvedFunction`. 

So far, Spark SQL does not handle star expansion when we use `star` in the DataFrame or DataSet functions. That is the reason I chose this title. Let me change it.

Actually, I am not sure if `CreateStruct` and `Count` are the only two functions that can accept `star`. Could you help me confirm it? Thanks!

**Comment by cloud-fan** [comment]
Actually we do handle stars in `CreateArray` and `CreateStruct`: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L440-L458, so what you are fixing is the nested `CreateStruct`, I think we should also add `CreateArray` too.

One of my concern is:  sometimes we check stars under `UnresolvedAlias` but sometimes also under `Alias`, it will be good if you can figure it out and make sure there is no missing case.

**Comment by gatorsmile** [comment]
uh, I see. The code you posted above is for `Project`. The error message in the original JIRA is for using `star` used in `Aggregate`. 

Yeah, we need a clean and complete fix for resolving star. Let me check if can move these into `expandStarExpressions`.

**Comment by gatorsmile** [comment]
@cloud-fan The latest commit separates star resolution from the reference resolution, since `ResolveReferences` becomes pretty long now. Could you help me check if the new changes cover all the cases that can accept star? Thank you! : )

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 4]
```diff
@@ -1834,6 +1834,8 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
       """.stripMargin).select($"r.*"),
       Row(3, 2) :: Nil)
 
+    assert(structDf.groupBy($"a").agg(min(struct($"record.*"))).first() == Row(3, Row(3, 1)))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: We should write a new test case to test `*` in `CreateStruct` and `CreateArray`, not just put in existing ones.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 5]
```diff
@@ -528,6 +528,14 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
     assert(e.getMessage.contains("cannot resolve 'c' given input columns: [a, b]"), e.getMessage)
   }
 
+  test("verify star in functions fail with a good error") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: Why put this test case in `DatasetSuite`?

**Comment by cloud-fan** [comment]
Overall LGTM except some comments about tests, thanks for working on it!

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala (hunk 5 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: This is copied from the example in the original JIRA. Let me move it. Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 4 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Sure, will do. Thanks!

**Comment by gatorsmile** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameComplexTypeSuite.scala [hunk 6]
```diff
@@ -30,6 +30,7 @@ class DataFrameComplexTypeSuite extends QueryTest with SharedSQLContext {
     val f = udf((a: String) => a)
     val df = sparkContext.parallelize(Seq((1, 1))).toDF("a", "b")
     df.select(struct($"a").as("s")).select(f($"s.a")).collect()
+    df.select(struct($"*").as("s")).select(f($"s.a")).collect()
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameComplexTypeSuite.scala]: not needed?

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameComplexTypeSuite.scala [hunk 7]
```diff
@@ -42,6 +43,7 @@ class DataFrameComplexTypeSuite extends QueryTest with SharedSQLContext {
     val f = udf((a: String) => a)
     val df = sparkContext.parallelize(Seq((1, 1))).toDF("a", "b")
     df.select(array($"a").as("s")).select(f(expr("s[0]"))).collect()
+    df.select(array($"*").as("s")).select(f(expr("s[0]"))).collect()
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameComplexTypeSuite.scala]: not needed?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 8]
```diff
@@ -1938,6 +1938,21 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
...
 
+  test("Star Expansion - CreateStruct and CreateArray") {
+    val structDf = testData2.select("a", "b").as("record")
+    // CreateStruct and CreateArray in aggregateExpressions
+    assert(structDf.groupBy($"a").agg(min(struct($"record.*"))).first() == Row(3, Row(3, 1)))
+    assert(structDf.groupBy($"a").agg(min(array($"record.*"))).first() == Row(3, Seq(3, 1)))
+
+    // CreateStruct and CreateArray in project list (unresolved alias)
+    assert(structDf.select(struct($"record.*")).first() == Row(Row(1, 1)))
+    assert(structDf.select(array($"record.*")).first().getAs[Seq[Int]](0) === Array(1, 1))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: nit: use `Seq(1, 1)` to match the type

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Done. : )

**Comment by gatorsmile** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 9]
```diff
@@ -350,28 +351,83 @@ class Analyzer(
...
+      // If the script transformation input contains Stars, expand it.
+      case t: ScriptTransformation if containsStar(t.input) =>
+        t.copy(
+          input = t.input.flatMap {
+            case s: Star => s.expand(t.child, resolver)
+            case o => o :: Nil
+          }
+        )
+      case g: Generate if containsStar(g.generator.children) =>
+        failAnalysis("Cannot explode *, explode can only be applied on a specific column.")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: just realized the error message is not clear enough, `Generate` is not always "explode"

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: do we have a test for this error message?

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: True. I moved this from another rule. I will check the coverage of test cases. Thanks!

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: We already have a test case: https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala#L181-L182

How about changing the message to `Invalid usage of '*' in explode/json_tuple/UDTF`? Thanks!

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: `explode/json_tuple/UDTF` LGTM

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Thanks! Let me change it now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 10]
```diff
@@ -369,28 +370,83 @@ class Analyzer(
...
+            UnresolvedAlias(child = expandStarExpression(ua.child, p.child)) :: Nil
+          case a @ Alias(_: UnresolvedFunction | _: CreateArray | _: CreateStruct, _) =>
+            Alias(child = expandStarExpression(a.child, p.child), a.name)(
+              isGenerated = a.isGenerated) :: Nil
+          case o => o :: Nil
+        }
+        Project(projectList = expanded, p.child)
+      // If the aggregate function argument contains Stars, expand it.
+      case a: Aggregate if containsStar(a.aggregateExpressions) =>
+        val expanded = a.aggregateExpressions.flatMap {
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Why `expandStarExpressions(a.aggregateExpression, a.child)` does not work?

**Comment by gatorsmile** [comment]
retest this please.

**Comment by cloud-fan** [comment]
cc @yhuai

**Comment by cloud-fan** [comment]
retest this please

**Comment by gatorsmile** [comment]
cc @yhuai : )

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
cc @yhuai @cloud-fan

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
cc @yhuai

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
cc @yhuai

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 11]
```diff
@@ -369,28 +370,83 @@ class Analyzer(
...
+            case s: Star => s.expand(child, resolver)
+            case o => o :: Nil
+          })
+        case c: CreateArray if containsStar(c.children) =>
+          c.copy(children = c.children.flatMap {
+            case s: Star => s.expand(child, resolver)
+            case o => o :: Nil
+          })
+        // count(*) has been replaced by count(1)
+        case o if containsStar(o.children) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: We can have a method:
```
private def mayContainsStar(expr: Expression): Boolean = expr.isInstnaceOf[UnresolvedFunction] || expr.isInstnaceOf[CreateStruct]...
```

then we can simplify this to:
```
expr.transformUp {
  case e if mayContainsStar(e) =>
    e.copy(children = ...)
}
```

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 12]
```diff
@@ -1934,6 +1934,21 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
...
+    val structDf = testData2.select("a", "b").as("record")
+    // CreateStruct and CreateArray in aggregateExpressions
+    assert(structDf.groupBy($"a").agg(min(struct($"record.*"))).first() == Row(3, Row(3, 1)))
+    assert(structDf.groupBy($"a").agg(min(array($"record.*"))).first() == Row(3, Seq(3, 1)))
+
+    // CreateStruct and CreateArray in project list (unresolved alias)
+    assert(structDf.select(struct($"record.*")).first() == Row(Row(1, 1)))
+    assert(structDf.select(array($"record.*")).first().getAs[Seq[Int]](0) === Seq(1, 1))
+
+    // CreateStruct and CreateArray in project list (alias)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: how about we add another case: `Generate` and `ScriptTransform`?

**Comment by cloud-fan** [comment]
Sorry for putting it here for such a long time, overall LGTM, will merge it after you address the new comments, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 11 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: That is a great idea! : )

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 12 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Sure, let me do it.

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Sure, let me do it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 11 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: That is a great idea! : )

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Tried it, but `copy` is unable to use here. When the type is `Expression` (abstract type), we are unable to use the `copy` function to change the `children`. In addition, `withNewChildren` requires the same number of children. Do you have any idea how to fix it? Thanks!

**Comment by gatorsmile** [comment]
@cloud-fan Thank you for your detailed reviews! I know all of you are very busy. Let me know if anything needs a change. Thanks again!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 13]
```diff
@@ -369,28 +370,83 @@ class Analyzer(
...
+
+      // If the projection list contains Stars, expand it.
+      case p: Project if containsStar(p.projectList) =>
+        val expanded = p.projectList.flatMap {
+          case s: Star => s.expand(p.child, resolver)
+          case ua @ UnresolvedAlias(_: UnresolvedFunction | _: CreateArray | _: CreateStruct, _) =>
+            UnresolvedAlias(child = expandStarExpression(ua.child, p.child)) :: Nil
+          case a @ Alias(_: UnresolvedFunction | _: CreateArray | _: CreateStruct, _) =>
+            Alias(child = expandStarExpression(a.child, p.child), a.name)(
+              isGenerated = a.isGenerated) :: Nil
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: We will lose qualifier here, how about `a.withNewChildren(expandStarExpression(a.child, p.child) :: Nil)`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 11 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: oh i see, I don't have a better idea, let's just keep it this way.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 14]
```diff
@@ -1934,6 +1934,43 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
...
+  test("Star Expansion - explode should fail with a meaningful message if it takes a star") {
+    val df = Seq(("1", "1,2"), ("2", "4"), ("3", "7,8,9")).toDF("prefix", "csv")
+    val e = intercept[AnalysisException] {
+      df.explode($"*") { case Row(prefix: String, csv: String) =>
+        csv.split(",").map(v => Tuple1(prefix + ":" + v)).toSeq
+      }.queryExecution.assertAnalyzed()
+    }
+    assert(e.getMessage.contains("Invalid usage of '*' in explode/json_tuple/UDTF"))
+
+    df.explode('prefix, 'csv) { case Row(prefix: String, csv: String) =>
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: put it in `checkAnswer` to make sure the execution is right.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 15]
```diff
@@ -737,6 +737,14 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
       .queryExecution.analyzed
   }
 
+  test("Star Expansion - script transform") {
+    val data = (1 to 100000).map { i => (i, i, i) }
+    data.toDF("d1", "d2", "d3").registerTempTable("script_trans")
+    assert(100000 ===
+      sql("SELECT TRANSFORM (*) USING 'cat' FROM script_trans")
+        .queryExecution.toRdd.count())
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: is it equal to `sql("...").count()`?

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Yeah, the results are the same. Let me simplify it. Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 14 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Sure, will do

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 13 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Yeah, a good catch!

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 16]
```diff
@@ -1934,6 +1934,50 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
     }
   }
 
+  test("Star Expansion - CreateStruct and CreateArray") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Why do we put these tests in `SQLQuerySuite`? It looks like they are mostly testing DF APIs.

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: True, let me move them to DataFrameSuite. Thanks!

**Comment by cloud-fan** [comment]
thanks! merging to master!

**Comment by davies** [comment]
@gatorsmile @cloud-fan This PR revert the change in https://github.com/apache/spark/pull/3674,  unfortunately the unit test in AnalysisSuite. This test break once we enforce max-iteration check in tests, see https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54090/testReport/org.apache.spark.sql.catalyst.analysis/AnalysisSuite/union_project__/

**Comment by davies** [comment]
@gatorsmile  This PR can't be easily reverted, so could you send a PR to fix it?

**Comment by davies** [comment]
I will fix this in #11828

**Pull request closed without merging**